### ISSUE TITLE
Provide fallback for systems that do not support X11 forwarding (getent: command not found)

### ISSUE THREAD

**Issue by frederickjh** [xxh/xxh-shell-fish#14]
I logged into a remote system and everything went well with the xxh-shell-fish install up to the point after the upload of fishportable. The welcome message for the system displayed then:

![image](https://user-images.githubusercontent.com/864159/100077100-3bbea080-2e42-11eb-8608-3621b6492562.png)

I checked and the executable `getent` is not available on this system. Not sure if this is what cause the following code to fail as well.

Any ideas on how to fix this or provide a fallback to something other than `getent` @anki-code ?

I am not totally sure I understand what line 122 of `entrypoint.sh` is doing.

**Comment by anki-code** [comment]
Hi @frederickjh!

> getent command not found

[This command](https://github.com/xxh/xxh-shell-fish/blob/87a788e1fc4a35bb2338b21bb2324db057447b1f/entrypoint.sh#L122) was added by #11 and the details could be found in https://github.com/xxh/xxh/issues/101

The command is to getting user HOME directory. It looks like we should have checking `getent` and have fallback if it does not exists. 

FYI @rdd2

**Comment by frederickjh** [comment]
So if I understand this correctly looking through the links you gave, It was added to allow X11 forwarding.
It my case X11 forwarding is blocked by the system as it is using chroot. 

So, what we need is a fallback for when a system does not support X11 forwarding.

Should we check if the command `getent` is present and if not skip this line?

I looked and found that if the environment variable **$DISPLAY** is not set that X11 forwarding will not work and that it will always be set correctly if forwarding is in place. Setting it manually will not work.

**Comment by tlaerm** [comment]
getent is used because it works with ldap setups (which I use a lot). The fallback should be  ```<pseudobash>cat /etc/shadow | grep $USER | cut ..</pseudobash>```

That would cover cases where the system has X11Forwarding but no getent.

If you want I'll whip something up and you can test.

**Comment by frederickjh** [comment]
@anki-code I am seeing that on some of the servers I currently have xxh-shell-fish installed on that if I did a reinstall it would fail. This change of adding the line that counts on `getent` being present on the remote system is a breaking change.

@rdd2 Yes, please do put a pull request together. I think it is good to have a fallback for `getent` but don't for get a fallback if no X11 forwarding is possible. I think the check here is to see if the environmental variable **$DISPLAY** is empty or set.

**Comment by frederickjh** [comment]
@rdd2 I am willing to test.  I have both chrooted server and ones with `getent`.

Not sure about your suggested fallback above for `getent`. It looks like you need super user permissions to access `/etc/shadow` on the systems I have tried.

**Comment by anki-code** [comment]
> @anki-code I am seeing that on some of the servers I currently have xxh-shell-fish installed on that if I did a reinstall it would fail. This change of adding the line that counts on getent being present on the remote system is a breaking change.

Yeah, we should think about xxh plugins versioning process some day.

> It looks like you need super user permissions to access /etc/shadow on the systems I have tried.

I think `/etc/passwd` was meant.

**Comment by frederickjh** [comment]
> I think `/etc/passwd` was meant.

In that case then:
```bash
cat /etc/passwd | grep $USER | cut -d: -f6
```
would return the users home directory. Is there a reason we can just get the home directory from **$HOME** in this case?

Since this is not the first time this project has run into issues with missing dependencies, I think we need to make a policy to check for dependencies before using them or if we have ones that we use multiple times we could check up front.

For a dependency called `foo` this could look like one of these:
```bash
command -v foo >/dev/null 2>&1 || { echo >&2 "I require foo but it's not installed.  Aborting."; exit 1; }

type foo >/dev/null 2>&1 || { echo >&2 "I require foo but it's not installed.  Aborting."; exit 1; }

hash foo 2>/dev/null || { echo >&2 "I require foo but it's not installed.  Aborting."; exit 1; }
```
I also have this in my notes about checking for programs:
> ... avoid **which**. Not only is it an external process you're launching for doing very little (meaning builtins like hash, type or command are way cheaper), you can also rely on the builtins to actually do what you want, while the effects of external commands can easily vary from system to system.

Of course we could also do something else other than aborting in the code example above, like falling back to another command, before aborting or asking them to file an issue against this project.

**Comment by anki-code** [comment]
> Is there a reason we can just get the home directory from $HOME in this case?

Good question. Probably we can save the `HOME` value [before modification](https://github.com/xxh/xxh-shell-fish/blob/ce5f0946b3d006e35137f674293385d8f305ceee/entrypoint.sh#L95-L106) and use it later for `XAUTHORITY`.

PR is welcome!

> Since this is not the first time this project has run into issues with missing dependencies

Yeah, so many systems with so many different environment. May be we should start writing cross platform functions i.e. `cross_platform_check_directory_exists`, `get_user_home` and others.

**Comment by frederickjh** [comment]
> Yeah, so many systems with so many different environment. May be we should start writing cross platform functions i.e. `cross_platform_check_directory_exists`, `get_user_home` and others.

I have to wonder if someone hasn't created a framework  for this that we could use.

**Comment by frederickjh** [comment]
I found that on the [Bashinator page](http://bashinator.org/) they also have a list of [other bash/shell frameworks](http://bashinator.org/#others). I have taken any time to see if one of these might be help for for this.

In any case this idea should probably be moved to an new issue, lest it be lost.

**Comment by anki-code** [comment]
I've added fallback for getent (https://github.com/xxh/xxh-shell-fish/commit/9d06d2b879d1adc104251709e2b28c9c4f1bbfe6) because new user faced with this error.
